File: android/app/src/main/java/com/dimento/beleganbei/DATEV.kt
package com.dimento.beleganbei; // replace your-apps-package-name with your app’s package name

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import com.dimento.beleganbei.development.MainApplication
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule
import datev.de.dcal.*


class DATEV(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    override fun getName() = "DATEV"

    fun DATEV(context: ReactApplicationContext) {
        // Set Auth Delegate
        DCAL.authDelegate = object : DCALapiDelegate {
            override fun dcalAuthStateChanged(newState: DCALState, error: DCALError?) {
                dcalAuthStateChanged(newState, error)
            }
        }

        // Set Session Delegate (= Progress Upload)
        DCAL.sessionDelegate = object : DCALSessionDelegate {
            override fun dcalSession(totalBytesExpectedToSend: Long, totalBytesSent: Long, totalBytesExpectedToReceive: Long, totalBytesReceived: Long) {
                dcalSession(totalBytesExpectedToSend, totalBytesSent, totalBytesExpectedToReceive, totalBytesReceived)
            }
        }
    }

    companion object {
        @JvmStatic fun handleIntent() {}

        @JvmStatic
        fun onNewIntent(intentString: Uri) {
            println("on NEw intent KT")
            DCAL.handleUrl(intentString)
        }
    }

    // Delegate for LOGIN/LOGOUT State, send to callback
    @ReactMethod
    private fun dcalAuthStateChanged(newState: DCALState, error: DCALError?) {
        Log.e("dcalAuthStateChanged", "")
        if (error != null) {
            println("dcalAuthStateChanged error $error")

            // Emit event to RN
            reactApplicationContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit("DATEV_AuthState_Error", error.errorMessage)
        } else {
            println("dcalAuthStateChanged newState $newState")

            // Emit event to RN
            reactApplicationContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit("DATEV_AuthState_Changed", isLoggedIn())
        }
    }

    // Delegate for UPLOAD Progress, send to callback
    @ReactMethod
    private fun dcalSession(totalBytesExpectedToSend: Long, totalBytesSent: Long, totalBytesExpectedToReceive: Long, totalBytesReceived: Long) {
        if (totalBytesExpectedToReceive > 0) {
            println("totalBytesExpectedToReceive $totalBytesExpectedToReceive")
            println("totalBytesReceived $totalBytesReceived")
            val percentage = (totalBytesReceived * 100.0 / totalBytesExpectedToReceive + 0.5).toInt()
            println("percentage % $percentage")
        }
        if (totalBytesExpectedToSend > 0) {
            val percentage = (totalBytesSent * 100.0 / totalBytesExpectedToSend + 0.5).toInt()
            println("percentage % $percentage")
        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun initialize(client_id: String, client_secret: String, scope: String, redirect_uri: String, use_sandbox: Boolean) {
        println("DATEV DCAL initialize")
        println("DATEV use_sandbox: $use_sandbox")
        println("DATEV client_id: $client_id")
        println("DATEV client_secret: $client_secret")
        println("DATEV scope: $scope")
        println("DATEV redirect_uri: $redirect_uri")

        val initResponse = DCAL.initialize(client_id, client_secret, scope, redirect_uri, use_sandbox, reactApplicationContext)
        println("DATEV initResponse: $initResponse")

        isInitialized();
        isLoggedIn();
        isSmartLoginAvailable();
    }

    @ReactMethod
    fun isSmartLoginAvailable() {
        println("DATEV isSmartLoginAvailable")
        val smartLoginAvailable = DCAL.isSmartLoginAvailable()

        // Emit event to RN
        reactApplicationContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit("DATEV_SMARTLOGIN_AVAILABLE", smartLoginAvailable);
    }

    @ReactMethod
    fun isInitialized() {
        println("DATEV isInitialized")
        val currentState = DCAL.getState()

        var result = false;
        if(!currentState.contains(State.uninitialized)) {
            result = currentState.contains(State.initialized);
        }

        // Emit event to RN
        reactApplicationContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit("DATEV_IS_INITIALIZED", result);
    }

    @ReactMethod
    fun isLoggedIn() {
        println("DATEV isLoggedIn")
        val currentState = DCAL.getState()
        val result = currentState.contains(State.loggedIn);

        // Emit event to RN
        reactApplicationContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit("DATEV_AuthState_Changed", result);
    }

    @ReactMethod
    fun requestLogin() {
        println("DATEV requestLogin")
        DCAL.requestLogin()
    }

    @ReactMethod
    fun requestLogout() {
        println("DATEV requestLogout")
        DCAL.requestLogout()

        Thread.sleep(1000L);
        isLoggedIn();
    }

    fun handleIntent(intentString: String, intentName: String) {
        var intentString = intentString
        println("DATEV handleIntent")
        intentString = intentString.replace("$intentName://datev?data=", "")

        // DEV HACK
        intentString = intentString.replace("beleganbeiopengrahl://datev?data=", "")
        intentString = intentString.replace("beleganbeiopendevelopment://datev?data=", "")
        // DEV HACK

        println("DATEV INTENT_STRING_FIXED $intentString")
        intentString = Uri.decode(intentString)
        println("DATEV INTENT_STRING_FIXED_2 $intentString")

        try {
            val datevCodeURL = Uri.parse(intentString)
            val DCAL = DCAL
            DCAL.handleUrl(datevCodeURL)

            Thread.sleep(1000L);
            isLoggedIn();

        } catch (e: Exception) {
            println("DATEV Exception DATEV Handle URL")
            println(e.localizedMessage)
        }
    }
}

File: ios/DATEV.swift
//
//  CustomMethods.swift
//  BelegAnbeiDevelopment
//
//  Created by Dieter Stratmann on 03.12.23.
//

import Foundation
import DCAL

@objc(DATEV) class DATEV: RCTEventEmitter, DCALapiDelegate, DCALSessionDelegate {
  
  @objc override static func requiresMainQueueSetup() -> Bool { return true }
  
  override func supportedEvents() -> [String]! {
    return ["DATEV_SMARTLOGIN_AVAILABLE", "DATEV_IS_INITIALIZED", "DATEV_AuthState_Changed", "DATEV_AuthState_Error"]
  }
  
  func dcalAuthStateChanged(newState state: DCAL.DCALState, withError error: Error?) {
    print("DCAL DELEGATE dcalAuthStateChanged")
    print("dcalAuthStateChanged state: \(state)")
    print("dcalAuthStateChanged error: \(String(describing: error))")
    
    let currentState = DCAL.getState()
    
    var isLN = false
    
    if currentState.contains(DCALState.loggedOut) {
      isLN = false
    } else {
      isLN = currentState.contains(DCALState.loggedIn)
    }
    
    DispatchQueue.main.async {
      self.sendEvent(withName: "DATEV_AuthState_Changed" , body: isLN)
    }
  }
  
  func dcalSession(totalBytesExpectedToSend: Int, totalBytesSent: Int, totalBytesExpectedToReceive: Int, didReceive: Int) {
    print("DCAL DELEGATE dcalSession")
    print("dcalSession totalBytesExpectedToSend: \(totalBytesExpectedToSend)")
    print("dcalSession totalBytesSent: \(totalBytesSent)")
    print("dcalSession totalBytesExpectedToReceive: \(totalBytesExpectedToReceive)")
    print("dcalSession didReceive: \(didReceive)")
  }
  
  @objc public func initialize(
    _ clientId: String,
    clientSecret: String,
    scope: String,
    redirectUri: String,
    useSandbox: Bool
  ) -> Void {
    print("initialize clientId: \(clientId)")
    print("initialize clientSecret: \(clientSecret)")
    print("initialize scope: \(scope)")
    print("initialize redirectUri: \(redirectUri)")
    print("initialize useSandbox: \(useSandbox)")
    
    DCAL.authDelegate = self;
    DCAL.sessionDelegate = self;

    let initResponse = DCAL.initialize(client_id: clientId, client_secret: clientSecret, scope: scope, redirect_URI: redirectUri, use_Sandbox: useSandbox)
    print("DATEV initResponse: \(initResponse)")
    
    isInitialized();
    isLoggedIn();
    isSmartLoginAvailable();
  }
  
  @objc public func handleURL(_ url: URL){
    print("handleURL url: \(url)")
    
    DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
      
      let loginHandled = DCAL.handleURL(url: url);
      
      print("handleURL loginHandled: \(loginHandled)")
      let currentState = DCAL.getState()
      
      var isLN = false
      
      if currentState.contains(DCALState.loggedOut) {
        isLN = false
      } else {
        isLN = currentState.contains(DCALState.loggedIn)
      }
      
      DispatchQueue.main.async {
        self.sendEvent(withName: "DATEV_AuthState_Changed" , body: isLN)
      }
      
      /*
      let currentState = DCAL.getState()
      print("handleURL currentState: \(currentState)")
      
      print("handleURL contains loggedOut: \(currentState.contains(DCALState.loggedOut))")
      print("handleURL contains loggedIn: \(currentState.contains(DCALState.loggedIn))")
      print("handleURL contains uninitialized: \(currentState.contains(DCALState.uninitialized))")
      print("handleURL contains initialized: \(currentState.contains(DCALState.initialized))")
      
      if currentState.contains(DCALState.loggedOut) {
        print("handleURL --> loggedOut: \(currentState.contains(DCALState.loggedOut))")
      } else {
        print("handleURL --> loggedIn: \(currentState.contains(DCALState.loggedIn))")
      }
      */
    }
  }
  
  @objc public func isSmartLoginAvailable() -> Void {
    print("isSmartLoginAvailable")
    
    DispatchQueue.main.async {
      let smartLoginAvailable = DCAL.isSmartLoginAvailable();
      self.sendEvent(withName: "DATEV_SMARTLOGIN_AVAILABLE" , body: smartLoginAvailable)
    }
  }
  
  @objc public func isInitialized() -> Void {
    print("isInitialized")

    let currentState = DCAL.getState()

    var result = false;
    if(!currentState.contains(DCALState.uninitialized)) {
        result = currentState.contains(DCALState.initialized);
    }
    
    DispatchQueue.main.async {
      self.sendEvent(withName: "DATEV_IS_INITIALIZED" , body: result)
    }
  }
  
  @objc public func isLoggedIn() -> Void {
    print("isLoggedIn")
    let currentState = DCAL.getState()
    let result = currentState.contains(DCALState.loggedIn);
    
    DispatchQueue.main.async {
      self.sendEvent(withName: "DATEV_AuthState_Changed" , body: result)
    }
  }
  
  @objc public func requestLogin() -> Void {
    print("requestLogin")
    DCAL.requestLogin()
  }
  
  @objc public func requestLogout() -> Void {
    print("requestLogout")
    //DCAL.revokeToken()
    DCAL.requestLogout()
   
    let clean = DCAL.clean()
    print("requestLogout --> clean: \(clean)")
  }
  
  static func stringify(json: Any, prettyPrinted: Bool = false) -> String {
      var options: JSONSerialization.WritingOptions = []
      if prettyPrinted {
        options = JSONSerialization.WritingOptions.prettyPrinted
      }

      do {
        let data = try JSONSerialization.data(withJSONObject: json, options: options)
        if let string = String(data: data, encoding: String.Encoding.utf8) {
          return string
        }
      } catch {
        print(error)
      }

      return ""
  }
}
